### ISSUE TITLE
Backward compatibility concerns with Softmax-like modules

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#77280]
### 🐛 Describe the bug

Some modules such as `Softmax`, `Softmin`, and `LogSoftmax` don't appear to call nn.Module superclass's __setstate__, for example, see https://github.com/pytorch/pytorch/blob/40576bceaf9dc2fc25796e8f9309d67a50b3f470/torch/nn/modules/activation.py#L1370. 

However, nn.Module __setstate__ is used to define certain attributes on loaded modules, i.e., if we add a new attribute `foo` to nn.Module, `__setstate__` is updated to set this attribute on nn.Modules that are being loaded from old PyTorch versions. In particular, this is nn.Module __setstate__:

https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1178

 This approach breaks with `Softmax` because the super call is not made. As a result, this seems like a backwards compatibility issue.

### Versions

main

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345

**Comment by jbschlosser** [comment]
Thanks for reporting the issue! Fix should be to add `super().__setstate__(state)` calls to the beginning of the `__setstate__` methods in each of `nn.Softmax`, `nn.Softmin`, and `nn.LogSoftmax`.

**Comment by rohan-varma** [comment]
We should probably also add documentation that if users override their own nn.Module setstate method, it needs to dispatch into the superclass, otherwise there will be BC issues

**Comment by jbschlosser** [comment]
> We should probably also add documentation that if users override their own nn.Module setstate method, it needs to dispatch into the superclass, otherwise there will be BC issues

Note that we generally discourage users from saving / loading entire modules, recommending operating with `state_dict`s instead. If they follow this advice, they won't have the BC issues. Documenting the workaround *might* implicitly add legitimacy to a non-blessed approach.

**Comment by rohan-varma** [comment]
fixed by https://github.com/pytorch/pytorch/pull/77392
